terminal/terminal.md

```diff
@@ -0,0 +1,417 @@
+
+- It's faster to operate on the terminal
+- Some data is not easily accessible through the GUI
+- Repetitive task can be automated through it
+- Makes difficult tasks easy
+
+And of course, when programming we will be using the terminal a lot, to install libraries, move across the filesystem, display changes in our files, push code to remote servers, etc.
+
+We are using BASH that is a command language or a command interpreter. BASH is a type of shell, a shell is an interface to the operating system, it passes commands to SO to carry out, there are other types of shells, you'll commonly see this two terms together
+
+## Hands on the terminal
+
+To start liking the terminal we need to know the basic commands, and keep learning accordingly to what we want to accomplish
+
+First let's open a terminal if we haven't done already. 
+```
+Daniels-MacBook-Air:apprentice danielvilla$ 
+```
+That is like called the **shell prompt** 
+
+Let's dissect the parts of the prompt:
+
+`Daniels-MacBook-Air` is the hostname
+
+`apprentice` is the current directory
+
+`danielvilla` is the name of the user
+
+`$` means you are a normal user
+
+the prompt is highly customizable changing the env variable **$PS1** but for now let's keep it the way it is
+
+Let's type whatever we want just to see what happen, let's experiment:
+```
+Daniels-MacBook-Air:apprentice danielvilla$ hey
+-bash: hey: command not found
+```
+As you can see the CLI *(command line interface)* is trying to resolve that name, it thinks it is a command, and as it wasn't found it return `command not found`, that tell you how to use it, and what it is expecting
+
+Now let's input a real command.
+
+Type `ls` to list all the files on the current directory. 
+
+```
+Daniels-MacBook-Air:apprentice danielvilla$ ls
+LICENSE   README.md
+```
+
+but wait how do we know in what directory we are? we use the command `pwd` that means *path of working directory*
+```
+Daniels-MacBook-Air:apprentice danielvilla$ pwd
+/Users/danielvilla/apprentice
+```
+Thre are two types of path, `absolute path` and `relative path` what pwd returns is the absolute path because it is NOT relative to the current directory we are on and it starts with **/**
+
+#### Navigation
+
+Now that we learned how to display the files on a directory (ls) let's learn how to move to another directory
+
+```
+danielvilla@Daniels-MacBook-Air:~$ cd docker
+danielvilla@Daniels-MacBook-Air:~/docker$ 
+```
+
+By the way, the **~** before the **$** sign indicates the home directory, all users have their own home directory, and is VERY IMPORTANT that you know that **/** indicates the root directory, off all the filesystem
+
+but now that we are on the docker directory how do we move back to the home directory, there are two ways:
+
+`danielvilla@Daniels-MacBook-Air:~/docker$ cd ..`
+
+and
+
+`danielvilla@Daniels-MacBook-Air:~/docker$ cd`
+
+**cd** without any arguments changes to your home directory no matter where you are
```

Comment:
+**cd** without any arguments changes to your home directory no matter where you are`.`